### PR TITLE
support hmis okta sign on

### PR DESCRIPTION
**Pull request by ttoomey**
### _Please squash-merge this PR_

Reference
[#183560765](https://www.pivotaltracker.com/story/show/183560765)
[hmis frontend #195](https://github.com/greenriver/hmis-frontend/pull/195)

Adjust okta integration to support login from the HMIS front end. The expectation is that the HMIS front end will use a distinct okta application.

The provider fields on user model are deprecated in favor of `oauth_identities` to allow a user to authenticate with more than one provider.

An "active_provider" cookie is stored so we know how to redirect the user after logout.

Per spec, added a settings endpoint `Hmis::AppSettingsController` for the HMIS front-end. For now this is just the okta endpoint, if enabled.

**testing:**
Tested against okta dev server. Requires creating a new okta "app":
```
Sign-in redirect URIs: https://hmis.dev.test/hmis/users/auth/okta/callback
Sign-out redirect URIs: https://hmis.dev.test
```
Then the client id and key should be set as env variables
```
HMIS_OKTA_CLIENT_ID=id_here
HMIS_OKTA_CLIENT_SECRET=secret_here
```

**misc:**
* use bang variant (`save!`) in in omniauth methods so save-failures are not silently ignored
* tidy logging in omniauth callback controller

### CODE REVIEW

File: app/controllers/users/omniauth_callbacks_controller.rb
Comment (eanders) [inline app/controllers/users/omniauth_callbacks_controller.rb]: Since these are used above, can we move them to the top of the file?

File: config/initializers/oauth2.rb
Comment (eanders) [inline config/initializers/oauth2.rb:9]: Thank you!

File: db/seeds.rb
Comment (eanders) [inline db/seeds.rb]: I don't _think_ we want to do this here without some guard on when it gets called.  `db:seed` is called on every deploy, and we don't want a test access group installed for production.
If this is used for testing, it could just get wrapped in an `if Rails.env.test?` or move to a factory.

Comment (ttoomey) [reply db/seeds.rb]: @eanders - Thanks, I didn't realize we were running this on deploy. I added some documentation for this to the top of the seeds file.

I noticed that `user.user_hmis_data_sources_roles` does not exist. Perhaps this line should just be removed?

Comment (eanders) [reply db/seeds.rb]: @gigxz any thoughts on `user.user_hmis_data_sources_roles`, I'm not familiar with that one.  Maybe that's obsolete after @ZachCMP 's recent upgrade to ACLs?  

Thanks @ttoomey, I think this looks good.

Comment (ttoomey) [reply db/seeds.rb]: thanks @eanders - my understanding is that I should rebase this into r70 once the sprockets issue is resolved.

Comment (eanders) [reply db/seeds.rb]: @ttoomey you can give https://github.com/greenriver/hmis-warehouse/pull/2910 a try and see if that solves it for you.  (It should.)  Otherwise, it's totally fine to wait until that gets merged into `release-70`

Comment (gigxz) [reply db/seeds.rb]: Yeah, I have some code in my local tasks for seeding the HMIS admin roles. I'll go back and update this file to be correct. We shouldn't be auto-assigning HMIS admin perms to a random user, so I'll remove that part.

File: app/controllers/users/omniauth_callbacks_controller.rb
Comment (gigxz) [inline app/controllers/users/omniauth_callbacks_controller.rb]: If we're calling this endpoint from the HMIS, we also need to add it to the rack cors config https://github.com/greenriver/hmis-warehouse/blob/release-70/config/initializers/cors.rb. Currently, we only allow requests to `/hmis/*` endpoints from the HMIS. I think it is safest to call out the specific endpoints, so `/users/auth/okta` instead of `/users/auth`.

For the reverse-proxying to work, you'll also need to make a copy of this block https://github.com/greenriver/hmis-frontend/blob/13f0af85c014f813e0d17baa898bb72c029aadf4/docker/templates/default.conf.template#L26-L36 for the new locations. Again, I think it's probably safest to proxy only the specific routes that we need.

Comment (ttoomey) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Thanks @gigxz! I've added the auth end point to the cors config. Good point about making the the endpoint more specific. I'll adjust the the nginx config template for the hmis-frontend too.

**Comment by ttoomey** [comment]
keeping as draft; will likely need to support multiple okta apps

File: config/initializers/omniauth.rb
Comment (ttoomey) [inline config/initializers/omniauth.rb]: TBD- need to find a way to test failure condition

File: app/controllers/concerns/omniauth_callback_behavior.rb
Comment (eanders) [inline app/controllers/concerns/omniauth_callback_behavior.rb]: This should be `user.my_root_path` for warehouse logins.

File: app/helpers/application_helper.rb
Comment (eanders) [inline app/helpers/application_helper.rb]: I'm probably missing something, but not _all_ HMIS installations will have an OKTA provider, only one at the moment, so that should probably be pulling from a database table or ENV.

File: app/models/omniauth_user_provisioner.rb
Comment (eanders) [inline app/models/omniauth_user_provisioner.rb]:
```suggestion
      now = Time.current
```

File: app/controllers/concerns/omniauth_callback_behavior.rb
Comment (ttoomey) [reply app/controllers/concerns/omniauth_callback_behavior.rb]: Looking at it again, this was only used by passthru, which was a vestige of devis/omniauthable. I don't think it's needed, removed this method

File: app/helpers/application_helper.rb
Comment (ttoomey) [reply app/helpers/application_helper.rb]: Thanks for spotting that, looks like I fat-fingered the wrong env variable there, my intention was `if ENV['OKTA_DOMAIN']`

This is is a replacement for  `resource_class.omniauth_providers` in the warehouse UI. If we need to move the provider config to the db we should check in on that since I've probably missed part of the spec

Comment (eanders) [reply app/helpers/application_helper.rb]: Nope, `OKTA_DOMAIN` makes way more sense to me.

**Comment by ttoomey** [comment]
Note, test failures seem to be related to `TodoOrDie::OverdueTodo` - I think they are unrelated to this PR

**Review by eanders — APPROVED** [review]
Looks good to me!  We should confirm on staging that it doesn't break the warehouse logins before merging.

**Comment by gigxz** [comment]
This and frontend look good to me! It did appear the base branches were somewhat out of sync - frontend was expecting a graphql attribute that wasn't present yet (2 prs merged by zach last week). Before deploying, can you merge in the latest release-70 to each?

**Review by eanders — APPROVED** [review]

File: spec/requests/sessions_controller_spec.rb
Comment (eanders) [inline spec/requests/sessions_controller_spec.rb]:
```suggestion
    it 'user sees an error' do
```

File: spec/requests/sessions_controller_spec.rb
Comment (eanders) [inline spec/requests/sessions_controller_spec.rb:148]: @ttoomey is this to work in conjunction with https://github.com/greenriver/hmis-warehouse/blob/e089fb53315c0b97bfb6a493231fd20115ad418e/.github/workflows/rails_tests.yml#L181-L183 or should we do something different/move this to a different test?  Maybe the old tests aren't relevant?

Comment (ttoomey) [reply spec/requests/sessions_controller_spec.rb:148]: I was just running the test locally where I can manage the env. I didn't realize we had  this one-off in CI already, that makes sense. Should I just add this test?

```diff
--- a/.github/workflows/rails_tests.yml
+++ b/.github/workflows/rails_tests.yml
@@ -180,7 +180,7 @@ jobs:
           LOG_LEVEL: INFO
         run: |
           bundle exec rspec --fail-fast=$MAX_FAILURES --color --format p -P "{spec,drivers}/**/*_spec.rb"
-          OKTA_DOMAIN=localhost OKTA_CLIENT_ID=x OKTA_CLIENT_SECRET=x bundle exec rspec --fail-fast=$MAX_FAILURES --color -fd spec/requests/omniauth_spec.rb # this one test needs a different ENV
+          OKTA_DOMAIN=localhost OKTA_CLIENT_ID=x OKTA_CLIENT_SECRET=x bundle exec rspec --fail-fast=$MAX_FAILURES --color -fd spec/requests/omniauth_spec.rb spec/requests/sessions_controller_spec.rb # this one test needs a different ENV
       - name: Run logging tests
```

II think the tests in session_conroller_spec and omniauth_spec are still relevant

Comment (eanders) [reply spec/requests/sessions_controller_spec.rb:148]: I think so.  Thanks!

File: config/environments/development.rb
Comment (gigxz) [inline config/environments/development.rb:131]: This is still needed for those of us who are developing the HMIS locally @ttoomey. Maybe you didn't need it because of your docker setup? Anyway, I'm adding it back

Comment (ttoomey) [reply config/environments/development.rb:131]: Sorry about that! I took it out to reproduce an issue I saw on QA related to rails origin forgery protection.

Comment (gigxz) [reply config/environments/development.rb:131]: Resolved with https://github.com/greenriver/hmis-frontend/pull/212 which makes [this function](https://apidock.com/rails/v5.1.7/ActionController/RequestForgeryProtection/valid_request_origin%3F) equally happy.

**Pull request merged**
